src/core/src/pass/stateful_to_stateless.cpp

```diff
@@ -0,0 +1,172 @@
+#include "openvino/op/gather.hpp"
+#include "openvino/op/read_value.hpp"
+#include "openvino/pass/manager.hpp"
+#include "transformations/utils/utils.hpp"
+
+using namespace ov::op;
+
+namespace {
+
+std::shared_ptr<ov::Node> set_name(std::shared_ptr<ov::Node> node, const std::string& name) {
+    // Set name for both node and output tensor (should be only one tensor, and any other names will be overriden by a
+    // given single name)
+    node->set_friendly_name(name);
+    OPENVINO_ASSERT(node->get_output_size() == 1);
+    node->get_output_tensor(0).set_names({name});
+    return node;
+}
+
+// Templated method that has the same effect as not templated `set_name` but saves Op type for convenient calls chaining
+template <typename T>
+inline std::shared_ptr<T> set_name(std::shared_ptr<T> node, const std::string& name) {
+    set_name(std::dynamic_pointer_cast<ov::Node>(node), name);
+    return node;
+}
+
+std::shared_ptr<v0::Parameter> get_parameter_by_tensor_name(const std::shared_ptr<ov::Model>& model,
+                                                            const std::string& name) {
+    for (const auto& param : model->get_parameters()) {
+        if (param->get_output_tensor(0).get_names().count(name))
+            return param;
+    }
+    return nullptr;  // nullptr and return type are only difference from ov::Model::input(name)
+}
+
+struct Variable {
+    struct Context {
+        // to hold compiled once regex for all Variable instances
+        const std::regex naming_convention =
+            std::regex(R"((past_key_values\.(\d+)\.(key|value))(present\.(\d+)\.(key|value)))");
+    };
+
+    Variable(const Context& context, const std::string& variable_name) : variable_name(variable_name) {
+        // Try to decode original naming of the corresponding input and output in the stateless model
+        std::smatch match;
+        if (std::regex_match(variable_name, match, context.naming_convention)) {
+            input_name = match[1].str();
+            output_name = match[4].str();
+            auto input_index = match[2].str();
+            auto output_index = match[5].str();
+            if (input_index == output_index && input_index.length() <= std::numeric_limits<int>::digits10) {
+                index = std::stoi(input_index) * 2 + int(match[3].str() == "value");  // order key before value
+            } else {
+                index = -1;
+            }
+        } else {
+            // Variable name doesn't follow the expected naming convention. It doens't prevent forming
+            // a correct stateless model but doesn't give a way to restore all names and inputs/outputs ordering
+            // accurately.
+            input_name = "input_restored." + variable_name;
+            output_name = "output_restored." + variable_name;
+            index = -1;
+        }
+    }
+
+    int index;                  // layer index, -1 means the index isn't known
+    std::string variable_name;  // original variable_id
```

Comment:
could you use "m_" prefix for the structure members?

Reply:
It will produce more "noise" in the code without a clear value. Use your regular conventions where it makes sense. Keep names simple if they are localized. It is just a simple struct that keeps a part of the code of the main function.